File: bindings/r/tests/testthat/test-spread.R
test_that("a sweep names every alternative it ran and takes no option to enable it", {
  standing <- quiet_trial()
  sweep <- pf_spread(
    standing,
    quantity = "system_weight_newtons",
    slot = "weighing",
    weighing = "bwepoch.fixed_window",
    onset = "onset.threshold.noise_relative",
    takeoff = "takeoff.threshold.absolute_force"
  )

  bindings <- pf_bindings()
  expect_identical(
    sweep[["combinations_run"]],
    sum(bindings[["slot"]] == "weighing")
  )
  expect_identical(
    sweep[["succeeded"]] + sweep[["failed"]],
    sweep[["combinations_run"]]
  )
  expect_identical(sweep[["quantity_key"]], "system_weight_newtons")
})

# A word the binding table holds no rule for. Written as `landing` this named a construct
# that later had two rules bound to it, so R built the axis and the engine refused it as one
# the request does not carry: a different sentence, and a green line turning red on a data
# edit nobody connected to it. The test below is the case `landing` moved into.
test_that("a step this build runs no rule for is refused with the ones it does", {
  step <- "not_a_step_this_build_runs"
  expect_false(step %in% pf_bindings()[["slot"]])

  condition <- tryCatch(
    pf_spread(
      quiet_trial(),
      quantity = "system_weight_newtons",
      slot = step,
      weighing = "bwepoch.fixed_window",
      onset = "onset.threshold.noise_relative",
      takeoff = "takeoff.threshold.absolute_force"
    ),
    plateforce_refusal = identity
  )

  expect_s3_class(condition, "plateforce_slot_offers_no_alternative")
  expect_identical(condition[["slot"]], step)
  expect_identical(
    conditionMessage(condition),
    paste0("this analysis runs no rule for ", step, ", so there is nothing to sweep")
  )
})

# A construct the build runs rules for and this request did not bind is not an axis: sweeping
# it would run a rule nobody chose. R passes the name through and the engine refuses it with
# the axes this request does carry, which is the answer the terminal and the notebook give.
test_that("a construct this request did not bind is refused with the axes it did", {
  bindings <- pf_bindings()
  landmarks <- c("weighing", "onset", "takeoff")
  # Two rules or more, or this surface refuses the step before the engine reads the axis and
  # the test would assert on a sentence from the wrong side of the wire.
  counted <- table(bindings[["slot"]])
  unbound <- setdiff(names(counted)[counted > 1], landmarks)
  skip_if(!length(unbound), "every construct with alternatives is one this request binds")

  condition <- tryCatch(
    pf_spread(
      quiet_trial(),
      quantity = "system_weight_newtons",
      slot = unbound[[1]],
      weighing = "bwepoch.fixed_window",
      onset = "onset.threshold.noise_relative",
      takeoff = "takeoff.threshold.absolute_force"
    ),
    plateforce_refusal = identity
  )

  expect_s3_class(condition, "plateforce_unknown_parameter")
  expect_true(all(landmarks %in% condition[["available"]]))
})

# The terminal and the notebook refuse a step this build runs one rule for, in one sentence.
# This surface let the name through, and the engine then refused it as a name that is not one
# of the axes a sweep can vary. A different fault from the one the caller has: it sends a
# reader looking for a typo or a binding they forgot, and never says the step is reached one
# way in this build.
#
# The step is read off the binding table rather than written here, so a rule added for it
# moves this test to a step that still holds one rather than leaving it asserting on a
# population that has moved on.
test_that("a slot this build runs one rule for is refused in the terminal's own sentence", {
  bindings <- pf_bindings()
  held_once <- names(Filter(function(count) count == 1L, table(bindings[["slot"]])))
  skip_if(!length(held_once), "this build runs two or more rules for every step")
  step <- held_once[[1]]

  condition <- tryCatch(
    pf_spread(
      quiet_trial(),
      quantity = "system_weight_newtons",
      slot = step,
      weighing = "bwepoch.fixed_window",
      onset = "onset.threshold.noise_relative",
      takeoff = "takeoff.threshold.absolute_force"
    ),
    plateforce_refusal = identity
  )

  expect_s3_class(condition, "plateforce_slot_offers_no_alternative")
  expect_identical(
    conditionMessage(condition),
    paste0("this analysis runs one rule for ", step, ", so there is nothing to sweep")
  )
})

test_that("several steps sweep every combination of their rules", {
  standing <- quiet_trial()
  bindings <- pf_bindings()
  steps <- c("weighing", "onset")

  sweep <- pf_spread(
    standing,
    quantity = "system_weight_newtons",
    slot = steps,
    weighing = "bwepoch.fixed_window",
    onset = "onset.threshold.noise_relative",
    takeoff = "takeoff.threshold.absolute_force"
  )

  # The product of the two, read off the binding table. Sweeping them one at a time reports
  # neither the widest disagreement nor the narrowest, which is why one call takes both.
  expected <- sum(bindings[["slot"]] == "weighing") * sum(bindings[["slot"]] == "onset")
  expect_gt(expected, sum(bindings[["slot"]] == "weighing"))
  expect_identical(sweep[["combinations_run"]], expected)
  expect_identical(length(sweep[["variants"]]), expected)
  expect_identical(sweep[["succeeded"]] + sweep[["failed"]], expected)
})

test_that("one step is one axis, whichever keyboard names it twice", {
  condition <- tryCatch(
    pf_spread(
      quiet_trial(),
      quantity = "system_weight_newtons",
      slot = c("onset", "onset"),
      weighing = "bwepoch.fixed_window",
      onset = "onset.threshold.noise_relative",
      takeoff = "takeoff.threshold.absolute_force"
    ),
    plateforce_refusal = identity
  )

  expect_s3_class(condition, "plateforce_value_not_accepted")
  expect_identical(
    conditionMessage(condition),
    "'onset' is named twice, and one step is one axis"
  )
})

test_that("a vector of method_ids describes one step, so several is refused", {
  condition <- tryCatch(
    pf_spread(
      quiet_trial(),
      quantity = "system_weight_newtons",
      slot = c("weighing", "onset"),
      method_ids = c("bwepoch.fixed_window", "bwepoch.manual_placement"),
      weighing = "bwepoch.fixed_window",
      onset = "onset.threshold.noise_relative",
      takeoff = "takeoff.threshold.absolute_force"
    ),
    plateforce_refusal = identity
  )

  expect_s3_class(condition, "plateforce_value_not_accepted")
  expect_identical(
    conditionMessage(condition),
    "a vector of method_ids describes one step, so name one step or key the ids by step"
  )
})

# Keyed by step, the ids say which rules belong to which, so named rules on two steps is one
# call. As a vector that question cannot be asked at all.
test_that("method_ids keyed by step names the rules for each of them", {
  bindings <- pf_bindings()
  weighing_rules <- bindings[["id"]][bindings[["slot"]] == "weighing"][1:2]
  onset_rules <- bindings[["id"]][bindings[["slot"]] == "onset"][1:2]

  sweep <- pf_spread(
    quiet_trial(),
    quantity = "system_weight_newtons",
    method_ids = list(weighing = weighing_rules, onset = onset_rules),
    weighing = "bwepoch.fixed_window",
    onset = "onset.threshold.noise_relative",
    takeoff = "takeoff.threshold.absolute_force"
  )

  expect_identical(sweep[["combinations_run"]], 4L)
  expect_identical(length(sweep[["axes_varied"]]), 2L)
})

# A length-one atomic vector is written as a scalar, so one value to sweep left here as a
# number where the engine reads a list of them. The fault came back naming a column of the
# request, which is a sentence about JSON rather than about the value the caller typed.
test_that("a setting sweep states its values as a list, one value or several", {
  standing <- quiet_trial()
  swept <- function(values) {
    pf_spread(
      standing,
      quantity = "system_weight_newtons",
      vary = list("weighing.duration" = values),
      weighing = "bwepoch.fixed_window",
      onset = "onset.threshold.noise_relative",
      takeoff = "takeoff.threshold.absolute_force"
    )
  }

  expect_identical(swept(c(0.5, 1.0))[["combinations_run"]], 2L)
  expect_identical(swept(0.5)[["combinations_run"]], 1L)
  # One combination holds no disagreement, so it publishes no spread. Reported as 0.0 it
  # said every alternative agreed, over a set nobody compared.
  expect_null(swept(0.5)[["spread_absolute"]])
  expect_false(is.null(swept(c(0.5, 1.0))[["spread_absolute"]]))
})

# The engine sweeps a rule and a value inside it on one request, and no surface could state
# that. On subject 01 trial 1 the six published values of onset.k move a jump height
# 0.01981 m against 0.01924 m for the five onset rules, so a reader asking about a number
# resting on both is asking one question.
test_that("the rules and a value inside them vary on one call", {
  standing <- quiet_trial()
  bindings <- pf_bindings()
  weighing_rules <- sum(bindings[["slot"]] == "weighing")

  both <- pf_spread(
    standing,
    quantity = "system_weight_newtons",
    slot = "weighing",
    vary = list("weighing.duration" = c(0.5, 1.0)),
    weighing = "bwepoch.fixed_window",
    onset = "onset.threshold.noise_relative",
    takeoff = "takeoff.threshold.absolute_force"
  )

  expect_identical(both[["combinations_run"]], weighing_rules * 2L)
  expect_identical(length(both[["axes_varied"]]), 2L)
  # The record names both, so a reader of the figure can see the whole set it came from.
  varied <- vapply(both[["axes_varied"]], function(axis) {
    if (is.null(axis[["parameter"]])) "rules" else axis[["parameter"]]
  }, character(1))
  expect_setequal(varied, c("rules", "duration"))
})

test_that("one setting is one axis, whichever way it is named twice", {
  condition <- tryCatch(
    pf_spread(
      quiet_trial(),
      quantity = "system_weight_newtons",
      vary = list("weighing.duration" = c(0.5, 0.5)),
      weighing = "bwepoch.fixed_window",
      onset = "onset.threshold.noise_relative",
      takeoff = "takeoff.threshold.absolute_force"
    ),
    plateforce_refusal = identity
  )

  expect_s3_class(condition, "plateforce_value_not_accepted")
  expect_identical(
    conditionMessage(condition),
    "weighing.duration names 0.5 twice, and one value is one variant"
  )
})

test_that("naming no step and no setting is refused rather than swept", {
  condition <- tryCatch(
    pf_spread(
      quiet_trial(),
      quantity = "system_weight_newtons",
      weighing = "bwepoch.fixed_window",
      onset = "onset.threshold.noise_relative",
      takeoff = "takeoff.threshold.absolute_force"
    ),
    plateforce_refusal = identity
  )

  expect_s3_class(condition, "plateforce_required_parameter_unstated")
  expect_identical(
    conditionMessage(condition),
    "no step and no setting were named, so there is nothing to sweep"
  )
})

test_that("a variant that declined is listed with its reason rather than dropped", {
  sweep <- pf_spread(
    quiet_trial(),
    quantity = "jump_height_from_takeoff_meters",
    slot = "onset",
    weighing = "bwepoch.fixed_window",
    onset = "onset.threshold.noise_relative",
    takeoff = "takeoff.threshold.absolute_force"
  )

  expect_identical(length(sweep[["variants"]]), sweep[["combinations_run"]])
  for (variant in sweep[["variants"]]) {
    expect_true(nzchar(variant[["label"]]))
    expect_true(!is.null(variant[["value"]]) || !is.null(variant[["failure_reason"]]))
  }
})

test_that("every onset rule this build runs computes, and they disagree by a real amount", {
  trace <- repository_trace()
  skip_if(is.null(trace), "the trial these fixtures were derived from is not here")
  trial <- pf_read_force_file(trace, sample_rate_hz = 1200, delimiter = "\t",
                              force_column = 0, sentinel_convention = "none")

  sweep <- pf_spread(
    trial,
    quantity = "jump_height_from_takeoff_meters",
    slot = "onset",
    weighing = "bwepoch.fixed_window",
    onset = "onset.threshold.noise_relative",
    takeoff = "takeoff.threshold.absolute_force"
  )

  values <- vapply(sweep[["variants"]], function(one) {
    if (is.null(one[["value"]])) NA_real_ else as.double(one[["value"]])
  }, numeric(1))

  expect_length(values, as.integer(fixture_field("onset-sweep.txt", "variants")))
  expect_identical(sweep[["failed"]], as.integer(fixture_field("onset-sweep.txt", "failed")))
  expect_equal(
    sweep[["spread_absolute"]],
    as.double(fixture_field("onset-sweep.txt", "spread_absolute_meters"))
  )

  # Which rule produced which number, rather than the summary alone. Held to four summary
  # figures the fixture certified the arithmetic and said nothing about the record, which is
  # the half of this product the arithmetic is in service of: a sweep that renamed every rule
  # and returned the same five numbers in the same order satisfied every line above.
  ran <- vapply(sweep[["variants"]], function(one) one[["settings"]][[1]][[2]], character(1))
  recorded <- strsplit(fixture_field("onset-sweep.txt", "method_ids"), ",", fixed = TRUE)[[1]]
  expect_identical(ran, recorded)
  expect_true(all(nzchar(recorded)))
  expect_identical(length(unique(recorded)), length(recorded))

  # Each rule against the number it produced, paired by position, so a reordering of the
  # document is a failure rather than a relabelling nobody sees.
  expect_equal(
    values,
    as.double(strsplit(fixture_field("onset-sweep.txt", "values_meters"), ",",
                       fixed = TRUE)[[1]])
  )

  # The registry these numbers came out of, as the registry declares itself. A record that
  # names no registry is a number a reader cannot place.
  expect_identical(
    if (is.null(sweep[["registry_declared_version"]])) "none" else
      sweep[["registry_declared_version"]],
    fixture_field("onset-sweep.txt", "registry_declared_version")
  )

  # No rule sits a long way from the rest, so the spread is five rules disagreeing rather
  # than one of them finding the wrong event. The two are the same number to a reader and
  # different findings to this project.
  furthest <- max(abs(values - stats::median(values, na.rm = TRUE)), na.rm = TRUE)
  expect_equal(furthest, as.double(fixture_field("onset-sweep.txt", "furthest_from_median_meters")))
  expect_lt(furthest, 0.1)
  expect_gt(sweep[["spread_absolute"]], 0)
})

File: crates/plateforce-python/tests/test_spread.py
"""The sweep, on a trace this file makes rather than one the checkout may or may not carry.

The spread is the measurement this software exists to publish: how far the choice of method
moves the number. A notebook could not compute it before this file existed, which left the
largest population of readers in this field able to run the analysis and unable to run the
argument.

Every test here reads a synthetic trace. They read a corpus trial until now, and every one of
them skipped wherever that trial is absent, which is every wheel this project ships: the
source distribution carries five of the workspace's crates and `plateforce-conformance` is
not among them, so `pytest {project}/crates/plateforce-python/tests` under cibuildwheel ran
eight skips and reported green. A skip is the one test result that cannot fail.

The question those eight could not answer, whether this surface's sweep is the terminal's, is
asked where it can be asked: `scripts/result-parity-requests.txt` names a `sweep` request, and
`scripts/result-parity.sh` holds this surface and the terminal to one committed record over
all 21 fields of a swept document on a real trial. That gate runs in a checkout, which is
where the trial is.

What a synthetic trace cannot carry, said once here rather than asserted weakly below. The
same sweep, the five published onset rules this build runs, spans 1.92 cm of jump height on
subject 01 trial 1 and 0.0038 cm on the trace `conftest.py` builds. So the magnitude is a
fact about real force data and no assertion here can hold it. It is held where real data is:
`tests/golden/result-parity-sweep.json` commits a wider sweep on that same trial, 75
combinations over all three landmark constructs spanning 3.11 cm, and the parity gate checks
it on every push.
"""

import numpy as np
import pytest

import plateforce as pf

from conftest import SAMPLE_RATE_HZ

# The three steps a jump height rests on, named here rather than defaulted inside the call,
# so a sweep over all three is asked for by this file rather than assumed.
LANDMARK_SLOTS = ["weighing", "onset", "takeoff"]

RULES = {
    "weighing": ("bwepoch.fixed_window", {"duration": 1.0}),
    "onset": ("onset.threshold.noise_relative", {"k": 5.0}),
    "takeoff": ("takeoff.threshold.absolute_force", {"threshold_n": 20.0}),
}

QUANTITY = "jump_height_from_takeoff_meters"


@pytest.fixture
def shipped_registry():
    """The registry this build embeds, which is the one a reader who pip installed has.

    Read rather than written, unlike the fixture registry the rest of the binding tests use:
    a sweep with no `method_ids` runs every rule the build's binding table holds for a slot,
    so the rules under test are the shipped ones either way and binding them from anywhere
    else would name ids the sweep does not reach.
    """
    return pf.Registry.load()


@pytest.fixture
def bound(shipped_registry):
    return {
        slot: shipped_registry.method(method_id).bind(**parameters)
        for slot, (method_id, parameters) in RULES.items()
    }


@pytest.fixture
def force_the_athlete_never_leaves(force_newtons):
    """The shared trace with its flight replaced by a settle back to standing.

    The shared trace ends in flight, so every combination swept over it places a takeoff and
    succeeds. A denominator asserted there is a denominator nothing was left out of, and the
    only trace on which a combination that produced nothing can be reached is one where the
    athlete stays on the plate.
    """
    flight_samples = int(0.5 * SAMPLE_RATE_HZ)
    on_the_plate = force_newtons[:-flight_samples]
    standing_newtons = float(on_the_plate[0])
    settling = np.linspace(float(on_the_plate[-1]), standing_newtons, flight_samples)
    return np.concatenate([on_the_plate, settling])


def sweep(trial, bound, slot, **stated):
    return pf.spread(
        trial,
        quantity=QUANTITY,
        slot=slot,
        weighing_epoch=bound["weighing"],
        onset=bound["onset"],
        takeoff=bound["takeoff"],
        **stated,
    )


def test_a_sweep_that_left_on_its_own_says_what_produced_it(trial, bound, shipped_registry):
    """A spread nested in an analysis inherits that result's identity. One that leaves alone
    carries its own, or a reader holding it cannot say which software or which registry
    answered, which is the whole claim this package makes.

    All four fields, because three of them were here and `registry_declared_version` was not:
    this surface published the digest and the caller's pin and not what the registry says
    about itself, while its own analysed result published all three and so did every other
    surface's sweep."""
    swept = sweep(trial, bound, "onset")
    assert swept.plateforce_version == pf.__version__
    assert swept.registry_digest == shipped_registry.digest
    assert swept.registry_declared_version == shipped_registry.declared_version
    # Null because nobody pinned one, which is the question this field answers and not the
    # one above it. The two were transposed on two surfaces for weeks.
    assert swept.registry_version is None
    assert swept.registry_version != swept.registry_declared_version


def test_the_method_choice_moves_the_number(trial, bound):
    """Five published onset rules, five different numbers, on one trace.

    How far apart they land is a fact about the data and it is measured on real data, in the
    committed sweep record this file's header names. What is asserted here is that choosing
    between them is a choice at all: five rules returning one number would mean the sweep
    varied nothing, and a spread of zero reads identically to a sweep that never ran."""
    swept = sweep(trial, bound, "onset")
    assert swept.succeeded == 5
    assert swept.failed == 0
    assert len({v.value for v in swept.variants}) == 5, "five onset rules, one number"
    assert swept.spread_absolute > 0
    assert swept.baseline_value == pytest.approx(0.053501495638726144)


def test_a_variant_carries_what_it_bound(trial, bound):
    swept = sweep(trial, bound, "onset")
    labels = {v.label for v in swept.variants}
    assert "onset onset.threshold.noise_relative" in labels
    for variant in swept.variants:
        assert variant.settings["onset"] in variant.method_ids
        assert variant.failure_reason is None


def test_sweeping_a_parameter_holds_the_rule(trial, bound):
    swept = sweep(trial, bound, "onset", parameter="k", values=[1.0, 5.0, 10.0])
    assert swept.combinations_run == 3
    assert {v.settings["k"] for v in swept.variants} == {"1", "5", "10"}
    assert all(v.method_ids[1] == RULES["onset"][0] for v in swept.variants)


def test_the_denominator_holds_the_combinations_that_produced_nothing(trial, bound_methods):
    """A combination that produced nothing is counted, listed, and says why.

    Swept over a takeoff threshold no sample of this trace falls below, beside one several
    samples do, so the failing half is reached. Asserted over one combination on a trace that
    places every landmark, `failed` is 0, `combinations_run == succeeded + failed` is
    arithmetic rather than a claim about this software, and a build that dropped a failing
    combination from the count or from the list reads exactly like one that kept it.

    A threshold of zero newtons is the choice a reader makes who wants the instant force
    reaches nothing at all. The plate reads zero in flight and never less, so the crossing the
    rule looks for is one this recording does not contain.
    """
    epoch, onset, takeoff = bound_methods
    swept = pf.spread(
        trial,
        quantity=QUANTITY,
        slot="takeoff",
        weighing_epoch=epoch,
        onset=onset,
        takeoff=takeoff,
        parameter="threshold_n",
        values=[20.0, 0.0],
    )
    assert swept.failed > 0, "nothing failed, so every claim below is about arithmetic"
    assert swept.succeeded > 0, "nothing succeeded, so the failure is the sweep and not the value"
    assert swept.combinations_requested == 2
    assert swept.combinations_run == 2
    assert swept.combinations_run == swept.succeeded + swept.failed
    assert len(swept.variants) == swept.combinations_run

    produced_nothing = [variant for variant in swept.variants if variant.value is None]
    assert len(produced_nothing) == swept.failed
    for variant in produced_nothing:
        assert variant.failure_reason is not None, f"{variant.settings} produced nothing in silence"


def test_a_sweep_that_computed_nothing_reports_its_denominator_and_no_width(
    trial, force_the_athlete_never_leaves, bound_methods
):
    """Every combination failed, and the answer is three of three rather than an empty one.

    No width beside it. A spread of zero over a set where nothing computed reads as the choice
    of method moving the number by nothing, which is the reading this file's sibling refusal
    exists to prevent, and it is the reading a reader cannot tell from a real agreement.
    """
    epoch, onset, takeoff = bound_methods
    swept = pf.spread(
        pf.Trial(force_the_athlete_never_leaves, sample_rate_hz=SAMPLE_RATE_HZ),
        quantity=QUANTITY,
        slot="onset",
        weighing_epoch=epoch,
        onset=onset,
        takeoff=takeoff,
        parameter="k",
        values=[1.0, 5.0, 10.0],
    )
    assert swept.succeeded == 0
    assert swept.failed == 3
    assert swept.combinations_run == 3
    assert len(swept.variants) == 3
    assert swept.spread_absolute is None
    assert swept.spread_percent_of_median is None
    assert swept.baseline_value is None
    for variant in swept.variants:
        assert variant.value is None
        assert variant.failure_reason is not None, f"{variant.settings} produced nothing in silence"

    # The control, and it can come back empty for the same reason the assertions above can: the
    # identical sweep on the shared trace, which ends in flight. A build that failed every
    # combination everywhere satisfies all of the above and fails here.
    with_flight = pf.spread(
        trial,
        quantity=QUANTITY,
        slot="onset",
        weighing_epoch=epoch,
        onset=onset,
        takeoff=takeoff,
        parameter="k",
        values=[1.0, 5.0, 10.0],
    )
    assert with_flight.succeeded == 3
    assert with_flight.failed == 0


def test_a_slot_this_build_runs_no_rule_for_is_refused(trial, bound):
    with pytest.raises(pf.MethodError) as raised:
        sweep(trial, bound, "not_a_step")
    assert "not_a_step" in str(raised.value)


def test_a_step_the_table_holds_one_rule_for_is_refused_in_the_terminals_words(trial, bound):
    """The terminal refuses `--slot time_to_takeoff`; this call ran it and reported zero.

    A step with one rule has nothing for that rule to be compared against, so the sweep ran a
    single variant and reported a spread of 0.0, which reads as the choice of method moving
    the number by nothing. That is the reading this software exists to prevent, and it was
    reachable from a notebook on six of the binding table's steps and from no terminal.

    `time_to_takeoff` is named rather than searched for, so this says which step it asks
    about. A second rule filed under that construct turns this red rather than quiet, which is
    the right way round: the step stops being an example and the test has to say so.

    The sentence is the terminal's, asserted whole rather than by a fragment of it, because
    the point is that a reader meets one wording whichever keyboard they are at. The
    terminal's half of the pair is `a_step_with_one_rule_is_refused_rather_than_dropped...`
    in `crates/plateforce-cli/tests/spread.rs`, which asserts these same words.
    """
    with pytest.raises(pf.MethodError) as raised:
        sweep(trial, bound, "time_to_takeoff")
    assert str(raised.value) == (
        "this analysis runs one rule for time_to_takeoff, so there is nothing to sweep"
    )

    # The control, and the half that says the floor is a floor rather than a wall: the five
    # onset rules are still swept, so a refusal that refused everything would not pass here.
    still_runs = sweep(trial, bound, "onset")
    assert still_runs.combinations_run == 5
    assert still_runs.spread_absolute > 0


def test_a_step_named_twice_is_one_axis_rather_than_a_sweep_squared(trial, bound):
    """One step is one axis, in the terminal's words.

    Named twice it was two axes: the same five onset rules ran 25 combinations, every one of
    them binding onset twice with the second binding winning, so twenty of the twenty-five
    repeated a rule and the denominator each figure was reported over counted a set nobody
    asked for. The terminal has refused this since the flag existed.
    """
    with pytest.raises(pf.MethodError) as raised:
        sweep(trial, bound, ["onset", "onset"])
    assert str(raised.value) == "'onset' is named twice, and one step is one axis"

    # Two different steps are two axes, which is the whole point of taking a list.
    both = sweep(trial, bound, ["onset", "takeoff"])
    assert both.combinations_run == 25


def test_a_parameter_and_several_slots_are_refused_together(trial, bound):
    """Each describes one slot, so the pair states a sweep nobody can mean."""
    with pytest.raises(pf.MethodError):
        sweep(trial, bound, LANDMARK_SLOTS, parameter="k", values=[1.0, 5.0])


def test_naming_no_slot_is_refused(trial, bound):
    with pytest.raises(pf.MethodError):
        sweep(trial, bound, [])


def test_the_cap_is_reported_rather_than_applied_silently(trial, bound):
    swept = sweep(trial, bound, LANDMARK_SLOTS, maximum_combinations=4)
    assert swept.capped is True
    assert swept.combinations_run == 4
    assert swept.combinations_requested > 4


def test_a_sweep_over_a_synthetic_trace_reports_its_unit(trial, bound_methods):
    epoch, onset, takeoff = bound_methods
    swept = pf.spread(
        trial,
        quantity="system_weight_newtons",
        slot="weighing",
        weighing_epoch=epoch,
        onset=onset,
        takeoff=takeoff,
        method_ids=[epoch.method_id],
    )
    assert swept.quantity_key == "system_weight_newtons"
    assert swept.unit == "newtons"
    assert swept.succeeded == 1
    assert swept.baseline_value == pytest.approx(
        swept.variants[0].value, rel=0, abs=0
    )
    assert len(swept) == 1
    assert SAMPLE_RATE_HZ == trial.sample_rate_hz
